libraries/AP_HAL_ChibiOS/AnalogIn.cpp

```diff
@@ -206,233 +236,389 @@ void AnalogSource::_add_value(float v, float vcc5V)
+        return pin_config[pin_index].channel;
+#if defined(HAL_ANALOG2_PINS)
+    case 1:
+        return pin_config_2[pin_index].channel;
+#endif
+#if defined(HAL_ANALOG3_PINS)
+    case 2:
+        return pin_config_3[pin_index].channel;
+#endif
+    };
+    osalDbgAssert(false, "invalid adc_index");
+    return 255;
+}
+
+uint8_t AnalogIn::get_analog_pin(uint8_t adc_index, uint8_t pin_index)
+{
+    switch(adc_index) {
+    case 0:
+        return pin_config[pin_index].analog_pin;
+#if defined(HAL_ANALOG2_PINS)
+    case 1:
+        return pin_config_2[pin_index].analog_pin;
+#endif
+#if defined(HAL_ANALOG3_PINS)
+    case 2:
+        return pin_config_3[pin_index].analog_pin;
+#endif
+    };
+    osalDbgAssert(false, "invalid adc_index");
+    return 255;
+}
+/*
+  return scaling
+ */
+float AnalogIn::get_pin_scaling(uint8_t adc_index, uint8_t pin_index)
+{
+    switch(adc_index) {
+    case 0:
+        return pin_config[pin_index].scaling;
+#if defined(HAL_ANALOG2_PINS)
+    case 1:
+        return pin_config_2[pin_index].scaling;
+#endif
+#if defined(HAL_ANALOG3_PINS)
+    case 2:
+        return pin_config_3[pin_index].scaling;
+#endif
+    };
+    osalDbgAssert(false, "invalid adc_index");
+    return 0;
+}
+
 /*
   callback from ADC driver when sample buffer is filled
  */
+#if HAL_WITH_MCU_MONITORING
+static uint16_t min_vrefint, max_vrefint;
+#endif
 void AnalogIn::adccallback(ADCDriver *adcp)
 {
-    const adcsample_t *buffer = samples;
-
-    stm32_cacheBufferInvalidate(buffer, sizeof(adcsample_t)*ADC_DMA_BUF_DEPTH*ADC_GRP1_NUM_CHANNELS);
+    uint8_t index = get_adc_index(adcp);
+    if (index >= HAL_NUM_ANALOG_INPUTS) {
+        return;
+    }
+    const uint16_t *buffer = (uint16_t*)samples[index];
+    uint8_t num_grp_channels = get_num_grp_channels(index);
+#if STM32_ADC_DUAL_MODE
+    if (index == 0) {
+        stm32_cacheBufferInvalidate(buffer, sizeof(uint16_t)*ADC_DMA_BUF_DEPTH*num_grp_channels*2);
+    } else
+#endif
+    {
+        stm32_cacheBufferInvalidate(buffer, sizeof(uint16_t)*ADC_DMA_BUF_DEPTH*num_grp_channels);
+    }
     for (uint8_t i = 0; i < ADC_DMA_BUF_DEPTH; i++) {
-        for (uint8_t j = 0; j < ADC_GRP1_NUM_CHANNELS; j++) {
-            sample_sum[j] += *buffer++;
+        for (uint8_t j = 0; j < num_grp_channels; j++) {
+            sample_sum[index][j] += *buffer;
+
+#if HAL_WITH_MCU_MONITORING
+            if (j == (num_grp_channels-1) && index == 2) {
+                // record min/max for MCU Vcc
+                if (min_vrefint == 0 ||
+                    min_vrefint > *buffer) {
+                    min_vrefint = *buffer;
+                }
+                if (max_vrefint == 0 ||
+                    max_vrefint < *buffer) {
+                    max_vrefint = *buffer;
+                }
+            }
+#endif
+            buffer++;
+#if STM32_ADC_DUAL_MODE
+            if (index == 0) {
+                // also sum the second ADC for dual mode
+                sample_sum[1][j] += *buffer;
```

Comment:
why [1] here instead of [0] when index is 0?

Reply:
also if index is 0 you are advancing buffer twice - is this correct?

Reply:
In Dual mode samples are alternate 16bit for ADC1 and then next 16bit for ADC2. This logic is to get the samples for each ADC separately